docs/readmes/lte/GTPU_Extension_Header.md

```diff
@@ -0,0 +1,139 @@
+                       actions.append(parser.OFPActionSetField(qfi = <qfi_value>)
+
+### RYU Controller
+
+Ryu is a Python library that provides an API wrapper for programming OVS.
+
+There are two parts to taking care of QFI functionality support:
+Nicira Extension Actions Structures (NXAction)
+Nicira Extended Match Structures (OFPMatch)
+
+### OVS Openflow
+
+Using the OpenFlow protocol, the controller can add, update, and delete flow entries in flow tables, both reactively (in response to packets) and proactively. Each flow table in the switch contains a set of flow entries; each flow entry consists of match fields, counters, and a set of instructions to apply to matching packets.
+
+### OVS Kernel
+
+For Egress flow, one needs to build the extension header and set the QFI value. It needs to be Passed down to kernel datapath for pushing it to GTP-U packets.
+For Ingress flow, extract the QFI value from skb of GTP header and set to the tunnel key.
+
+## High Level Call Flow
+
+Call flow for QFI set in OVS.
+
+```mermaid
+sequenceDiagram
+    CPE->>MME(AMF): PDU Session Establishment Request
+    MME(AMF)->>SESSIOND: SetSMFContext
+    SESSIOND->>PIPELINED: SetSMFSession(GTP)
+    PIPELINED->>RYU_CONTROLLER: Set QFI value in match and action criteria
+    RYU_CONTROLLER->>OPENFLOW: NXM_NX_QFI
+    OPENFLOW->>OVS_DATAPATH: MFF_QFI and set tunnel QFI
+    CPE->>OVS_DATAPATH: GTP Traffic for uplink/downlink with QFI value
+```
+
+## Common Issues and Troubleshooting
+
+1. First check GTP table entries using `sudo ovs-ofctl -O OpenFlow13 dump-flows gtp_br0 table=0`
+2. Check whether GTP port `2152` is listening or not by using `sudo netstat -a|grep 2152`
+   If it is not listening, check the `OVS` status using `sudo ovs-vsctl show`
+3. `ovs-dpctl` dumps the current state of kernel flow table. This table reflects active connections in the system, so it needs running traffic between UE andinternet server.
+4. If the below error is observed by running `sudo ovs-vsctl show`
+
+   Port gtp0
+   Interface gtp0
+   type: gtp
+   options: {key=flow, remote_ip=flow}
+   error: "could not add network device gtp0 to ofproto (Address family not supported by protocol)"
+
+   Then GTP tunnel type is changed to `gtpu` in `/etc/network/interfaces.d/gtp` file.
+5. GTP kernel module is included as part of OVS module. So no need to insert gtp.ko
+
+6. To setup DEV environment, run the following command on the magma-dev VM
+   `sudo bash ~/magma/third_party/gtp_ovs/ovs-gtp-patches/2.15/dev.sh setup`
+
+7. To Run OVS GTP tests on OVS kernel datapath:
+   `sudo bash ~/magma/third_party/gtp_ovs/ovs-gtp-patches/2.15/dev.sh build_test`
+
+8. OVS debug logging can be dynamically enabled by `sudo ovs-appctl vlog/set dbg`
+   For a specific module,
+   `sudo ovs-appctl vlog/set netdev dbg`
+   `sudo ovs-appctl vlog/set ofproto dbg`
+   `sudo ovs-appctl vlog/set vswitchd dbg`
+   `sudo ovs-appctl vlog/set dpif dbg`
+
+9. Debug the traffic issues in fastpath

```

Comment:
Currently this is rendered like so: "Debug the traffic issues in fastpath Enable the OVS debug logging Check the logs using sudo dmesg". Perhaps it could be rewritten as: "To debug the traffic issues in fastpath, enable the OVS debug logging and check the logs using `sudo dmesg`".

Reply:
Fixed it.